src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi

```diff
@@ -437,48 +437,79 @@ cdef class Metadatum:
 cdef class _MetadataIterator:
 
   cdef size_t i
-  cdef Metadata metadata
+  cdef size_t _length
+  cdef object _metadatum_indexable
 
-  def __cinit__(self, Metadata metadata not None):
+  def __cinit__(self, length, metadatum_indexable):
+    self._length = length
+    self._metadatum_indexable = metadatum_indexable
     self.i = 0
-    self.metadata = metadata
 
   def __iter__(self):
     return self
 
   def __next__(self):
-    if self.i < len(self.metadata):
-      result = self.metadata[self.i]
+    if self.i < self._length:
+      result = self._metadatum_indexable[self.i]
       self.i = self.i + 1
       return result
     else:
       raise StopIteration
 
 
+# TODO(https://github.com/grpc/grpc/issues/7950): Eliminate this; just use an
+# ordinary sequence of pairs of bytestrings all the way down to the
+# grpc_call_start_batch call.
 cdef class Metadata:
+  """Metadata being passed from application to core."""
 
   def __cinit__(self, metadata_iterable):
+    metadata_sequence = tuple(metadata_iterable)
+    cdef size_t count = len(metadata_sequence)
     with nogil:
       grpc_init()
-      grpc_metadata_array_init(&self.c_metadata_array)
-    metadata = list(metadata_iterable)
-    for metadatum in metadata:
-      if not isinstance(metadatum, Metadatum):
-        raise TypeError("expected list of Metadatum")
-    self.c_metadata_array.count = len(metadata)
-    self.c_metadata_array.capacity = len(metadata)
+      self.c_metadata = <grpc_metadata *>gpr_malloc(
+          count * sizeof(grpc_metadata))
+      self.c_count = count
+    for index, metadatum in enumerate(metadata_sequence):
+      self.c_metadata[index].key = grpc_slice_copy(
+          (<Metadatum>metadatum).c_metadata.key)
+      self.c_metadata[index].value = grpc_slice_copy(
+          (<Metadatum>metadatum).c_metadata.value)
+
+  def __dealloc__(self):
+    with nogil:
+      for index in range(self.c_count):
+        grpc_slice_unref(self.c_metadata[index].key)
+        grpc_slice_unref(self.c_metadata[index].value)
+      gpr_free(self.c_metadata)
+      grpc_shutdown()
+
+  def __len__(self):
+    return self.c_count
+
+  def __getitem__(self, size_t index):
+    if index < self.c_count:
+      key = _slice_bytes(self.c_metadata[index].key)
+      value = _slice_bytes(self.c_metadata[index].value)
+      return Metadatum(key, value)
+    else:
+      raise IndexError()
+
+  def __iter__(self):
+    return _MetadataIterator(self.c_count, self)
+
+
+cdef class MetadataArray:
```

Comment:
As I understand it, this is this is a Python wrapper for the array we pass to c-core that gets returned filled with the metadata.  Do we need to unreference the returned slices at all?  Or does that happen somewhere else?

It looks like ```grpc_metadata_array_destroy()``` just frees the array.

Reply:
Good question; I think this was at the center of what we were getting wrong before. Based on the specifications [here](https://github.com/grpc/grpc/blob/7f6a27a4d3675adc0de263d291346aab2f46e733/include/grpc/impl/codegen/grpc_types.h#L530-L536) and [here](https://github.com/grpc/grpc/blob/7f6a27a4d3675adc0de263d291346aab2f46e733/include/grpc/impl/codegen/grpc_types.h#L544-L550), this structure deliberately _does not_ unref the `grpc_slice`s it accesses.

A consequence of this is that instances of this class are only legitimate for use while their associated call is not-yet-destroyed, and a consequence of that are the `tuple` calls added in `CompletionQueue._interpret_event` and `Operation.received_metadata`.

Does the `TODO` added to `Operation.received_metadata` help enough, or should I add more text calling out this pitfall?

Reply:
I'm satisfied based on the description in ```grpc_types.h```